No more dep in makefile/build - long live dep through vendor/

ldemailly opened the pull request:
Follow up on having vendor/ checked in (as submodule)

We don't need to run a bunch of dep commands anymore, or build/find dep for that matter

This should shave some time from the build, not counting the simpler Makefile and script

We can later make new targets to automate
https://github.com/istio/istio/wiki/Vendor-FAQ#how-do-i-add--change-a-dependency
(for instance we can follow up with #3679)

ldemailly commented on Makefile:
```diff
@@ -215,38 +213,14 @@ depend: init | $(ISTIO_OUT)
-	unset GOOS && CGO_ENABLED=1 go get -u github.com/golang/dep/cmd/dep
-
 ${GOLINT}:
 	unset GOOS && CGO_ENABLED=1 go get -u github.com/golang/lint/golint
```
btw we should probably not get golint that way either - I think bin/linter.sh use an image ...

mattdelco replied:
There were some other linter targets that I think have since been removed so this target (and the GOLINT variable) might now be unused (though I didn't search the other .mk files that are imported by this one).

ldemailly replied:
indeed
```
ldemailly-macbookpro:istio ldemailly$ git grep GOLINT
Makefile:GOLINT   := $(shell which golint 2>/dev/null || echo "${ISTIO_BIN}/golint" )
Makefile:${GOLINT}:
```
will drop it but I want to see all builds green first before touching the pr again and cancelling the inflight ones

ldemailly commented:
circle builds and the part of prow that needs deps are happy with this pr 🎉 
wtb "/lgtm"

ldemailly commented:
before (on master, dependencies circleci build) : 1min50s init step
![screen shot 2018-02-21 at 9 50 53 pm](https://user-images.githubusercontent.com/3664595/36522454-79200852-1751-11e8-890c-306ee1a537a5.png)

after: 0min02sec !

![screen shot 2018-02-21 at 9 51 26 pm](https://user-images.githubusercontent.com/3664595/36522449-74454dce-1751-11e8-9f28-b311a0055bd4.png)

costinm commented on Makefile:
```diff
@@ -215,38 +213,14 @@ depend: init | $(ISTIO_OUT)
-	$(Q) ${DEP} ensure -update
-
-# Generates the current status of the dependencies. Does not update the deps.
-depend.status:
```
Please keep those targets, maybe redirect the output to the OUT directory. It's something we may want to include somewhere in the release or artifacts.

ldemailly replied:
i'd prefer to not keep all the DEP variable and build magic just to have a pretty graph target that probably noone uses (and whoever did can type "dep status -dot | dot -T png > dep.png" ?)

I think that should go in follow ups like #3679

ldemailly replied:
I'm adding it back because when I tried it complained about input-digest mismatch
so it's useful for that

mattdelco commented on Makefile:
```diff
@@ -215,38 +213,14 @@ depend: init | $(ISTIO_OUT)
-
-# Target to update the Gopkg.lock with latest versions.
-# Should be run when adding any new dependency and periodically.
-depend.update: ${DEP} ; $(info $(H) ensuring dependencies are up to date...)
```
A few lines above (where github won't let me add a comment, sigh), there's a reference to "must manually call dep.update" that probably meant "depend.update", which is now deleted in this change.

ldemailly replied:
updating that part

ldemailly commented:
Gopkg.lock is [checked in](https://github.com/istio/istio/blob/master/Gopkg.lock), that's not changing (though #3679 would be good to have to make sure it doesn't go stale/out of sync with vendor-istio)

also ideally there would be some sort of commit hook / alert on changes to that file and vendor/

ldemailly commented:
For the record I was trying to add a working "depend.update" because of some other pr issue in #oncall but was having random errors again like:
```
$ dep ensure
Solving failure: No versions of gopkg.in/validator.v2 met constraints:
	v2: Could not introduce gopkg.in/validator.v2@v2, as its subpackage gopkg.in/validator.v2 does not contain usable Go code (*build.NoGoError).. (Package is required by (root).)
	v2: Could not introduce gopkg.in/validator.v2@v2, as its subpackage gopkg.in/validator.v2 does not contain usable Go code (*build.NoGoError).. (Package is required by (root).)
```

I fixed them by putting deletion of the .lock in that target and removed a bunch of files in my go path it took a very long time but it worked. definitely something we should do on a nightly or weekly build rather than by hand

guptasu commented:
/lgtm

ldemailly commented:
matt, yes I am trying to do only minor surgery in this PR on our makefile but I agree with you.

nmittler commented:
@ldemailly I was just chatting with @costinm and he's thinking that this PR incorrectly removed the `--update` from the `depend.update` target. I'm going to put a PR together to revert that target.
